compiler/rustc_const_eval/src/transform/check_consts/check.rs

```diff
@@ -752,142 +750,61 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
-
-                    let implsrc = {
-                        let infcx = tcx.infer_ctxt().build();
-                        let mut selcx = SelectionContext::new(&infcx);
-                        selcx.select(&obligation)
-                    };
-
-                    match implsrc {
-                        Ok(Some(ImplSource::Param(_))) if tcx.features().effects => {
-                            debug!(
-                                "const_trait_impl: provided {:?} via where-clause in {:?}",
-                                trait_ref, param_env
-                            );
-                            return;
-                        }
-                        // Closure: Fn{Once|Mut}
-                        Ok(Some(ImplSource::Builtin(BuiltinImplSource::Misc, _)))
-                            if trait_ref.self_ty().is_closure()
-                                && tcx.fn_trait_kind_from_def_id(trait_id).is_some() =>
-                        {
-                            let ty::Closure(closure_def_id, fn_args) = *trait_ref.self_ty().kind()
-                            else {
-                                unreachable!()
-                            };
-                            if !tcx.is_const_fn_raw(closure_def_id) {
-                                self.check_op(ops::FnCallNonConst {
-                                    caller,
-                                    callee,
-                                    args: fn_args,
-                                    span: *fn_span,
-                                    call_source: *call_source,
-                                    feature: None,
-                                });
-
-                                return;
-                            }
-                        }
-                        Ok(Some(ImplSource::UserDefined(data))) => {
+                    // trait method calls are only permitted when `effects` is enabled.
+                    // we don't error, since that is handled by typeck. We try to resolve
+                    // the trait into the concrete method, and uses that for const stability
+                    // checks.
+                    // FIXME(effects) we might consider moving const stability checks to typeck as well.
+                    if tcx.features().effects {
+                        is_trait = true;
+                        let trait_ref = TraitRef::from_method(tcx, trait_id, fn_args);
+                        let obligation =
+                            Obligation::new(tcx, ObligationCause::dummy(), param_env, trait_ref);
+
+                        let implsrc = {
+                            let infcx = tcx.infer_ctxt().build();
+                            let mut selcx = SelectionContext::new(&infcx);
+                            selcx.select(&obligation)
+                        };
+                        if let Ok(Some(ImplSource::UserDefined(data))) = implsrc {
```

Comment:
Instead of just matching on `ImplSource::UserDefined`, I feel like we should just directly call `Instance::resolve` here:

```
 if let Ok(Some(func)) = instance
    && let InstanceDef::Item(def) = func.def
{
    callee = def;
}
```

Reply:
Maybe also then comment that this is just necessary to select a more specific item for const stability. If the instance-resolve fails, then we'll look at the const stability of the trait item, rather than a more specific impl item.

Reply:
done

Reply:
Why are we still using select here? I don't think it's needed with the `Instance::resolve` call.

Reply:
Perhaps I misread your previous comment. I will fix in a moment.